README.md
# PDBHTMLReport

This is a [KBase](https://kbase.us) module generated by the [KBase Software Development Kit (SDK)](https://github.com/kbase/kb_sdk).

You will need to have the SDK installed to use this module. [Learn more about the SDK and how to use it](https://kbase.github.io/kb_sdk_docs/).

You can also learn more about the apps implemented in this module from its [catalog page](https://narrative.kbase.us/#catalog/modules/PDBHTMLReport) or its [spec file]($module_name.spec).

# Setup and test

Add your KBase developer token to `test_local/test.cfg` and run the following:

```bash
$ make
$ kb-sdk test
```

After making any additional changes to this repo, run `kb-sdk test` again to verify that everything still works.

# Installation from another module

To use this code in another SDK module, call `kb-sdk install PDBHTMLReport` in the other module's root directory.

# Help

You may find the answers to your questions in our [FAQ](https://kbase.github.io/kb_sdk_docs/references/questions_and_answers.html) or [Troubleshooting Guide](https://kbase.github.io/kb_sdk_docs/references/troubleshooting.html).


RELEASE_NOTES.md
# PDBHTMLReport release notes
=========================================

0.0.0
-----
* Module created by kb-sdk init


test/PDBHTMLReport_server_test.py
# -*- coding: utf-8 -*-
import os
import time
import unittest
from configparser import ConfigParser

from PDBHTMLReport.PDBHTMLReportImpl import PDBHTMLReport
from PDBHTMLReport.PDBHTMLReportServer import MethodContext
from PDBHTMLReport.authclient import KBaseAuth as _KBaseAuth

from installed_clients.WorkspaceClient import Workspace


class PDBHTMLReportTest(unittest.TestCase):

    @classmethod
    def setUpClass(cls):
        token = os.environ.get('KB_AUTH_TOKEN', None)
        config_file = os.environ.get('KB_DEPLOYMENT_CONFIG', None)
        cls.cfg = {}
        config = ConfigParser()
        config.read(config_file)
        for nameval in config.items('PDBHTMLReport'):
            cls.cfg[nameval[0]] = nameval[1]
        # Getting username from Auth profile for token
        authServiceUrl = cls.cfg['auth-service-url']
        auth_client = _KBaseAuth(authServiceUrl)
        user_id = auth_client.get_user(token)
        # WARNING: don't call any logging methods on the context object,
        # it'll result in a NoneType error
        cls.ctx = MethodContext(None)
        cls.ctx.update({'token': token,
                        'user_id': user_id,
                        'provenance': [
                            {'service': 'PDBHTMLReport',
                             'method': 'please_never_use_it_in_production',
                             'method_params': []
                             }],
                        'authenticated': 1})
        cls.wsURL = cls.cfg['workspace-url']
        cls.wsClient = Workspace(cls.wsURL)
        cls.serviceImpl = PDBHTMLReport(cls.cfg)
        cls.scratch = cls.cfg['scratch']
        cls.callback_url = os.environ['SDK_CALLBACK_URL']
        suffix = int(time.time() * 1000)
        cls.wsName = "test_ContigFilter_" + str(suffix)
        ret = cls.wsClient.create_workspace({'workspace': cls.wsName})  # noqa

    @classmethod
    def tearDownClass(cls):
        if hasattr(cls, 'wsName'):
            cls.wsClient.delete_workspace({'workspace': cls.wsName})
            print('Test workspace was deleted')

    # NOTE: According to Python unittest naming rules test method names should start from 'test'. # noqa
    def test_your_method(self):
        # Prepare test objects in workspace if needed using
        # self.getWsClient().save_objects({'workspace': self.getWsName(),
        #                                  'objects': []})
        #
        # Run your method by
        # ret = self.getImpl().your_method(self.getContext(), parameters...)
        #
        # Check returned data with
        # self.assertEqual(ret[...], ...) or other unittest methods
        pdb_infos = [{
            'structure_name': '6TUK',
            'file_extension': '.pdb',
            'narrative_id': 63679,
            'genome_name': 'MLuteus_ATCC_49442',
            'feature_id': 'MLuteus_masurca_RAST.CDS.133',
            'is_model': 1,
            'from_rcsb': 1,
            'file_path': os.path.join('/kb/module/test/data', '6TUK.pdb.gz'),
            'genome_ref': '63679/38/1',  # for appdev
            # 'genome_ref': '107138/2/1',  # for prod
            'feature_type': 'gene',
            'sequence_identities': '68.93%',
            'chain_ids': 'Model 1.Chain B',
            'model_ids': '0',
            'exact_matches': '0',
            'scratch_path': os.path.join('/kb/module/test/data', '6TUK.pdb.gz')
        }, {
            'structure_name': 'MLuteus_AlphaFold_3483',
            'file_extension': '.pdb',
            'narrative_id': 63679,
            'genome_name': 'MLuteus_ATCC_49442',
            'feature_id': 'MLuteus_masurca_RAST.CDS.3483',
            'is_model': 1,
            'from_rcsb': 0,
            'file_path': os.path.join('/kb/module/test/data', 'MLuteus_AlphaFold_3483.pdb'),
            'genome_ref': '63679/38/1',  # for appdev
            # 'genome_ref': '107138/2/1',  # for prod
            'feature_type': 'gene',
            'sequence_identities': '99.99%',
            'chain_ids': 'Model 1.Chain B',
            'model_ids': '0',
            'exact_matches': '1',
            'scratch_path': os.path.join('/kb/module/test/data', 'MLuteus_AlphaFold_3483.pdb')
        }]
        ret = self.serviceImpl.run_PDBHTMLReport(self.ctx, pdb_infos)
        self.assertTrue(ret[0]['report_html'])
        self.assertTrue(ret[0]['doms'])
        print(ret[0]['report_html'])


lib/PDBHTMLReport/PDBHTMLReportImpl.py
# -*- coding: utf-8 -*-
#BEGIN_HEADER
import logging
import os

from PDBHTMLReport.Utils.PDBReportUtil import PDBReportUtil
#END_HEADER


class PDBHTMLReport:
    '''
    Module Name:
    PDBHTMLReport

    Module Description:
    A KBase module: PDBHTMLReport
    '''

    ######## WARNING FOR GEVENT USERS ####### noqa
    # Since asynchronous IO can lead to methods - even the same method -
    # interrupting each other, you must be *very* careful when using global
    # state. A method could easily clobber the state set by another while
    # the latter method is running.
    ######################################### noqa
    VERSION = "0.0.1"
    GIT_URL = ""
    GIT_COMMIT_HASH = "95ab4907e56dcb18abaa0e53959a77c98814a444"

    #BEGIN_CLASS_HEADER
    '''
    PDBHTMLReport will take a metadata file from the staging area and output an html report
    '''
    #END_CLASS_HEADER

    # config contains contents of config file in a hash or None if it couldn't
    # be found
    def __init__(self, config):
        #BEGIN_CONSTRUCTOR
        self.config = config
        self.config['SDK_CALLBACK_URL'] = os.environ['SDK_CALLBACK_URL']
        self.config['KB_AUTH_TOKEN'] = os.environ['KB_AUTH_TOKEN']

        logging.basicConfig(format='%(created)s %(levelname)s: %(message)s',
                            level=logging.INFO)
        #END_CONSTRUCTOR
        pass


    def run_PDBHTMLReport(self, ctx, params):
        """
        This function accepts PDBHTMLInput as input parameters and returns an output in a PDBHTMLReport
        :param params: instance of type "PDBHTMLReportInput" -> structure:
           parameter "pdb_infos" of list of type "PDBHInfo" (Input/Output of
           run_PDBHTMLReport structure_name: name of the protein structure
           rcsb_id: The structure id for RCSB database file_extension: the
           file extension of the structure, default to 'pdb' from_rcsb:
           indicator if the strcture is from the RCSB Database, default to 0
           is_model: indicator if the strcture is from computational
           modeling, default to 1 narrative_id: id of a KBase narrative
           genome_name: name of a KBase genome object genome_ref: name of a
           KBase genome object reference feature_id: id of a KBase feature
           object feature_type: id of a KBase feature object's type, default
           to 'gene' file_path: path to the pdb_file file of the pdb
           structure stratch_path: path on the shared folder where the
           structure file resides, default to file_path sequence_identities:
           sequence identities matched chain_ids: protein chain ids
           model_ids: model ids exact_matches: a string comma seperated '0'
           and '1' indicating an exact match not found ('0') or found ('1')
           for the structure's proteins with a given KBase genome feature
           @optional) -> structure: parameter "structure_name" of String,
           parameter "rcsb_id" of String, parameter "file_extension" of
           String, parameter "narrative_id" of String, parameter
           "genome_name" of String, parameter "genome_ref" of String,
           parameter "feature_id" of String, parameter "feature_type" of
           String, parameter "file_path" of String, parameter "scratch_path"
           of String, parameter "sequence_identities" of String, parameter
           "chain_ids" of String, parameter "model_ids" of String, parameter
           "exact_matches" of String, parameter "is_model" of type "boolean"
           (A boolean - 0 for false, 1 for true. @range (0, 1)), parameter
           "from_rcsb" of type "boolean" (A boolean - 0 for false, 1 for
           true. @range (0, 1))
        :returns: instance of type "PDBHTMLReportOutput" (report_html: a
           complete html document that can be opened by any browser doms: an
           UnspecifiedObject with DOM html segments that can be used by other
           apps) -> structure: parameter "report_html" of String, parameter
           "doms" of unspecified object
        """
        # ctx is the context object
        # return variables are: output
        #BEGIN run_PDBHTMLReport
        self.config['USER_ID'] = ctx['user_id']
        self.pdb_util = PDBReportUtil(self.config)

        output = self.pdb_util.generate_html_report(params)

        #END run_PDBHTMLReport

        # At some point might do deeper type checking...
        if not isinstance(output, dict):
            raise ValueError('Method run_PDBHTMLReport return value ' +
                             'output is not type dict as required.')
        # return the results
        return [output]
    def status(self, ctx):
        #BEGIN_STATUS
        returnVal = {'state': "OK",
                     'message': "",
                     'version': self.VERSION,
                     'git_url': self.GIT_URL,
                     'git_commit_hash': self.GIT_COMMIT_HASH}
        #END_STATUS
        return [returnVal]


lib/PDBHTMLReport/Utils/PDBReportUtil.py
import logging
import os
import sys
import uuid
import shutil
from urllib.parse import urlparse

from installed_clients.DataFileUtilClient import DataFileUtil
from installed_clients.WorkspaceClient import Workspace


class PDBReportUtil:

    def _validate_html_report_params(self, params):
        """
            _validate_html_report_params:
                validates params passed to generate_html_report method
        """
        # check for required parameters
        for p in ['pdb_infos', 'workspace_name']:
            if p not in params:
                raise ValueError(f'Parameter "{p}" is required, but missing')

    def _config_viewer(self, div_id):
        """
            _config_viewer: write the mol* viewer configurations
        """
        return (f'<script type="text/javascript">'
                f'molstar.Viewer.create("{div_id}", {{'
                f'layoutIsExpanded: false,'
                f'layoutShowControls: true,'
                f'layoutShowRemoteState: false,'
                f'layoutShowSequence: true,'
                f'layoutShowLog: true,'
                f'layoutShowLeftPanel: true,'
                f'viewportShowExpand: true,'
                f'viewportShowSelectionMode: true,'
                f'viewportShowAnimation: true,'
                f'collapseLeftPanel: true,'
                f'}}).then(viewer => {{')

    def _save_structure_file(self, dest_dir, pdb):
        """
            _save_structure_file: Place a copy of the structure file to the dest_dir
        """
        file_path = pdb['file_path']
        pdb_file_path = pdb['scratch_path']  # this is the scratch path for this pdb file
        new_pdb_path = os.path.join(dest_dir, os.path.basename(file_path))
        if not pdb.get('from_rcsb', False):
            shutil.copy(pdb_file_path, new_pdb_path)

    def _write_viewer_content_single(self, output_dir, pdb_infos):
        """
            _write_viewer_content_single: write the mol* viewer html content to fill in the
                                        subtabcontent and replace the string
                                        <!--replace subtab content single-->
                                        in the templatefile 'pdb_report_template.html'
        """
        viewer_content = ''
        pdb_index = 0

        for pdb in pdb_infos:
            if pdb.get('from_rcsb', False):
                continue
            file_path = pdb['file_path']
            file_ext = pdb['file_extension']
            if file_ext == 'cif':
                file_ext = 'mmcif'
            base_filename = os.path.basename(file_path)

            struct_nm = pdb['structure_name'].upper()
            div_id = 'struct' + str(pdb_index + 1)

            sub_div = (f'<div id="{struct_nm}" class="subtabcontent">'
                       f'<h2>{struct_nm}</h2>'
                       f'<div id="{div_id}" class="app"></div>')

            script_content = self._config_viewer(div_id)
            script_content += (f'viewer.loadStructureFromUrl("./{base_filename}", '
                               f'"{file_ext}", false, {{ representationParams: '
                               f'{{ theme: {{ globalName: "operator-name" }} }} }});'
                               '});')
            script_content += '</script>'

            sub_div += script_content
            sub_div += '</div>'
            viewer_content += sub_div
            pdb_index += 1

        return viewer_content

    def _write_viewer_content_multi(self, output_dir, pdb_infos):
        """
            _write_viewer_content_multi: write the mol* viewer html content to fill in the
                                         subtabcontent and replace the string
                                         <!--replace StructureViewer subtabs-->
                                         in the templatefile 'pdb_report_template.html'
        """
        div_id = 'app_all'
        viewer_script = self._config_viewer(div_id)

        default_click = ''
        viewer_tabs = '<div class="tab">'
        pre_loads = ''

        for pdb in pdb_infos:
            if pdb.get('from_rcsb', False):
                continue
            struct_nm = pdb['structure_name'].upper()
            if not default_click:
                default_click = f'{struct_nm}_sub'
            file_path = pdb['file_path']
            file_ext = pdb['file_extension']
            if file_ext == 'cif':
                file_ext = 'mmcif'

            viewer_tabs += (f'<button id="{struct_nm}_sub" '
                            f'class="subtablinks" onclick="openSubTab(event, this)">'
                            f'{struct_nm}</button>')

            pre_loads += (f'viewer.loadStructureFromUrl("./{os.path.basename(file_path)}", '
                          f'"{file_ext}", false, {{ representationParams: '
                          f'{{ theme: {{ globalName: "operator-name" }} }} }});'
                          '});')
        viewer_tabs += '</div>'

        # insert the structure file for preloading
        if pre_loads:
            viewer_script += pre_loads
            viewer_script += '</script>'
            viewer_script += '</div>'
        else:
            viewer_script = ''

        return viewer_tabs, viewer_script, default_click

    def _write_structure_info(self, output_dir, pdb_infos):
        """
            _write_structure_info: write the batch uploaded structure info to replace the string
                                   '<!--replace uploaded pdbs tbody-->' in the tboday tag of the
                                   jQuery DataTable in the template file 'pdb_report_template.html'
        """

        tbody_html = ''
        srv_domain = urlparse(self.shock_url).netloc  # parse url to get the domain portion
        srv_base_url = f'https://{srv_domain}'
        if srv_base_url == 'https://kbase.us':
            srv_base_url = 'https://narrative.kbase.us'

        logging.info(f"Get the url for building the anchor's href: {srv_base_url}")

        for pdb in pdb_infos:
            self._save_structure_file(output_dir, pdb)

            if pdb.get('structure_name', None):
                struct_id = pdb['structure_name'].upper()
            elif pdb.get('rcsb_id', None):
                struct_id = pdb['rcsb_id'].upper()
            else:
                struct_id = 'Unknown strtucture name'

            genome_name = pdb['genome_name']
            genome_ref = pdb['genome_ref']
            feat_id = pdb['feature_id']

            pdb_chains = []
            seq_idens = []
            if pdb.get('chain_ids', None):
                pdb_chains = pdb['chain_ids']
            if pdb.get('sequence_identities', None):
                seq_idens = pdb['sequence_identities']

            # Start writing the tbody content of the table
            tbody_html += '<tr>'
            frm_rcsb = pdb.get('from_rcsb', False)
            if not frm_rcsb:
                tbody_html += (f'<td><div class="subtablinks" '
                               f'onclick="openSubTab(event, this, false)" '
                               f'style="cursor:pointer;color:blue;text-decoration:underline;" '
                               f'title="Click to see in mol*">{struct_id}</div></td>')
            else:
                tbody_html += (f'<td><a href="https://www.rcsb.org/structure/{struct_id}"'
                               f'style="cursor: pointer;" target="_blank" '
                               f'title="3D Structure Viewer">{struct_id}</a></td>')

            tbody_html += (f'<td><a href="{srv_base_url}/#dataview/{genome_ref}"'
                           f' target="_blank">{genome_name}</a></td><td>{feat_id}</td>')
            tbody_html += f'<td>{pdb_chains} </td>'
            tbody_html += f'<td>{seq_idens}</td>'
            tbody_html += '</tr>'

        return tbody_html

    def __init__(self, config):
        self.callback_url = config['SDK_CALLBACK_URL']
        self.scratch = config['scratch']
        self.token = config['KB_AUTH_TOKEN']
        self.dfu = DataFileUtil(self.callback_url)
        self.ws_client = Workspace(config['workspace-url'])
        self.shock_url = config['shock-url']
        self.download_dir = None
        self.__baseDownloadUrl = 'https://files.rcsb.org/download'

    # def generate_html_report(self, pdb_infos):
    def generate_html_report(self, pdb_infos):
        """
            generate_html_report: generates the HTML for the report by 
            1. call _validate_html_report_params to validate input params
            2. generate a report in html

            input pdb_infos is actually pdb_infos, which is a list of of the following structure,
            e.g.,
            pdb_infos = [{
                'structure_name': '6TUK',
                'file_extension': '.pdb',
                'narrative_id': 63679,
                'genome_name': 'MLuteus_ATCC_49442',
                'feature_id': 'MLuteus_masurca_RAST.CDS.133',
                'is_model': 1,
                'from_rcsb': 1,
                'file_path': os.path.join('/kb/module/test/data', '6TUK.pdb.gz'),
                'genome_ref': '63679/38/1',
                'feature_type': 'gene',
                'sequence_identities': '68.93%',
                'chain_ids': 'Model 1.Chain B',
                'model_ids': '0',
                'exact_matches': '0',
                'scratch_path': os.path.join('/kb/module/test/data', '6TUK.pdb.gz')
            }]
            return: an html string
        """
        # Make report directory for writing and copying over uploaded pdb files
        output_directory = os.path.join(self.scratch, str(uuid.uuid4()))
        os.mkdir(output_directory)

        tbody_html = self._write_structure_info(output_directory, pdb_infos)

        dir_name = os.path.dirname(__file__)
        report_template_file = os.path.join(dir_name, 'templates', 'pdb_report_template.html')
        single_viewer = self._write_viewer_content_single(output_directory, pdb_infos)
        viewer_tabs, multi_viewer, default_click = self._write_viewer_content_multi(
                                                                      output_directory,
                                                                      pdb_infos)
        # Note: multi_viewer is not used because we decide to leave out to speed up the report
        # loading. In case we want to add that tab, would only need to add inside the following
        # `with` block a line of code of:
        #   html_report = html_report.replace('<!--replace subtab content multi-->', multi_viewer)
        # In the meantime, the viewer_tabs should be concatenated with another line like:
        # viewer_tabs += ('<button id="AllStructures_sub" class="subtablinks" '
        #                 'onclick="openSubTab(event, this)">ALL STRUCTURES</button>')
        with open(report_template_file, 'r') as report_template_pt:
            # Fetch & fill in detailed info into template HTML
            if tbody_html:
                html_report = report_template_pt.read()\
                    .replace('<!--replace uploaded pdbs tbody-->', tbody_html)
            if viewer_tabs:
                html_report = html_report\
                    .replace('<!--replace StructureViewer subtabs-->', viewer_tabs)
            if single_viewer:
                html_report = html_report\
                    .replace('<!--replace subtab content single-->', single_viewer)
            html_report = html_report\
                .replace('document.getElementById("AllStructures_sub").click()',
                         f'document.getElementById("{default_click}").click()')
            html_report = html_report.replace('\\', '')
            html_report = html_report.replace('\n', '')

        return {'report_html': html_report,
                'doms': {
                    'table_content': tbody_html,
                    'viewer_tabs': viewer_tabs,
                    'single_structures': single_viewer,
                    'all_structures': multi_viewer
                }}
